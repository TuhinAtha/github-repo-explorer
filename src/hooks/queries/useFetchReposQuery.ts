import { useQuery } from "@tanstack/react-query";
import queryKeys from "./QueryKeys";
import Repo from "../../models/Repo";
import { AxiosError } from "axios";
import GitHubAxiosInstance from "../../GitHubAxiosInstance";

const DEFAULT_PER_PAGE = 5

export type ReposQueryProps = {
  username: string,
  per_page?: number
}

const useFetchReposQuery = (props: ReposQueryProps) => {
  const { username = '', per_page = DEFAULT_PER_PAGE } = props

  return useQuery<Repo[], AxiosError>({
    queryKey: [queryKeys.USER_REOPS, { ...props }],
    queryFn: async () => {
      const response = await GitHubAxiosInstance.get(`/users/${username}/repos?per_page=${per_page}`, {
        headers: {
          'Accept': 'application/vnd.github.v3+json',
          'X-GitHub-Api-Version': '2022-11-28'
        }
      })
      return response.data
    },
    enabled: Boolean(username)
  })
}

export default useFetchReposQuery;
