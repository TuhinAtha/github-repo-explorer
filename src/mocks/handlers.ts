import { rest } from 'msw'

export const handlers = [rest.get(
  process.env.REACT_APP_GITHUB_BASE_URL + '/search/users',
  (req, res, ctx) => {
    return res(
      ctx.json([{
        login: 'username1'
      }])
    )
  }),
rest.get(
  process.env.REACT_APP_GITHUB_BASE_URL + '/users/:username/repos',
  (req, res, ctx) => {
    return res(
      ctx.json([{
        name: 'repo1',
        description: 'repo desctipiton 1',
        stargazers_count: 100
      }])
    )
  }
)]
