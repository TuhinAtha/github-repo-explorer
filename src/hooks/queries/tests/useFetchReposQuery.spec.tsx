import { renderHook, waitFor } from '@testing-library/react'
import useFetchReposQuery from '../useFetchReposQuery'
import { ReactQueryClientProviderWrapper } from '../../../mocks/ReactQueryClientProvider'

describe('Hooks -> useFetchReposQuery', () => {
  it('Should work fine', async () => {
    const { result } = renderHook(
      () =>
        useFetchReposQuery({
          username: 'username1',
        }),
      { wrapper: ReactQueryClientProviderWrapper }
    )

    await waitFor(() => expect(result.current.isSuccess).toBe(true));
    expect(result.current.data).toMatchSnapshot()
  })
})
