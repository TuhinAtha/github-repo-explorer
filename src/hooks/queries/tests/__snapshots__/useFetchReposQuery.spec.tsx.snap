// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hooks -> useFetchReposQuery Should work fine 1`] = `
Array [
  Object {
    "description": "repo desctipiton 1",
    "name": "repo1",
    "stargazers_count": 100,
  },
]
`;
